src/v/model/tests/timeout_adl.cc

```diff
@@ -35,3 +40,52 @@ BOOST_AUTO_TEST_CASE(test_timeout_clock_is_ms_2) {
     auto res = reflection::adl<model::timeout_clock::duration>{}.from(p);
     BOOST_REQUIRE_EQUAL(res, d);
 };
+
+// In https://github.com/redpanda-data/redpanda/pull/5051
+// the implementation of adl<model::timeout_clock::duration> was refactored
+// This is roughly the old implementation.
+struct adl_legacy_duration {
+    using rep = model::timeout_clock::rep;
+    using duration = model::timeout_clock::duration;
+
+    void to(iobuf& out, duration dur) {
+        // This is a clang bug that cause ss::cpu_to_le to become ambiguous
+        // because rep has type of long long
+        // adl<rep>{}.to(out, dur.count());
+        reflection::adl<uint64_t>{}.to(
+          out,
+          std::chrono::duration_cast<std::chrono::milliseconds>(dur).count());
+    }
+    duration from(iobuf_parser& in) {
+        // This is a clang bug that cause ss::cpu_to_le to become ambiguous
+        // because rep has type of long long
+        // auto rp = adl<rep>{}.from(in);
+        auto rp = reflection::adl<uint64_t>{}.from(in);
```

Comment:
think we are relying on some undefined behavior / compiler optimization here.. rp should overflow for -ve durations? (testcase: -16ms)

Reply:
No I don't think so:
* Overflow is defined for unsigned.
* 2's complement is standardised representation.

These are all the same, `fffffffffffffff0`
```

int main() {
    int64_t negative{-16};

    std::cout << std::hex << negative << '\n';
    std::cout << std::hex  << uint64_t(negative) << '\n'; // not brace initialised, narrowing
    std::cout << std::hex << *reinterpret_cast<const uint64_t* const>(&negative) << '\n';
}
```

Reply:
Ya that makes sense with 2's complement repro, for some reason I thought that was UB and not allowed.